compiler-rt/lib/ctx_profile/tests/RootAutoDetectorTest.cpp

```diff
@@ -0,0 +1,134 @@
+#include "../RootAutoDetector.h"
+#include "sanitizer_common/sanitizer_array_ref.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using namespace __ctx_profile;
+using ::testing::IsEmpty;
+using ::testing::SizeIs;
+class MockCallsiteTree final : public PerThreadCallsiteTrie {
+  // Return the first multiple of 100.
+  uptr getFctStartAddr(uptr CallsiteAddress) const override {
+    return (CallsiteAddress / 100) * 100;
+  }
+};
+
+// Utility for describing a preorder traversal. By default it captures a value -
+// the value of a node. Implicitly nodes are expected to have 1 child. If they
+// have none, we place a Marker::term and if they have more than one, we place a
+// Marker::split(nr_of_children)
+// For example, using lists: (1 (2 3) (4 (5 6)))
+// is a list of markers:
+// 1, split(2), 2, term, 3, term, 4, split(2), 5, term, 6, term
+class Marker {
+  enum class Kind { End, Value, Split };
+  const uptr Value;
+  const Kind K;
+  Marker(uptr V, Kind S) : Value(V), K(S) {}
+
+public:
+  Marker(uptr V) : Marker(V, Kind::Value) {}
+
+  static Marker split(uptr V) { return Marker(V, Kind::Split); }
+  static Marker term() { return Marker(0, Kind::End); }
+
+  bool isSplit() const { return K == Kind::Split; }
+  bool isTerm() const { return K == Kind::End; }
+  bool isVal() const { return K == Kind::Value; }
+
+  bool operator==(const Marker &M) const {
+    return Value == M.Value && K == M.K;
+  }
+};
+
+void popAndCheck(ArrayRef<Marker> &Preorder, Marker M) {
+  ASSERT_FALSE(Preorder.empty());
+  ASSERT_EQ(Preorder[0], M);
+  Preorder = Preorder.drop_front();
+}
+
+void checkSameImpl(const Trie &T, ArrayRef<Marker> &Preorder) {
+  popAndCheck(Preorder, T.address());
+
+  if (T.children().size() == 0) {
+    popAndCheck(Preorder, Marker::term());
+    return;
+  }
+
+  if (T.children().size() > 1)
+    popAndCheck(Preorder, Marker::split(T.children().size()));
+
+  T.children().forEach([&](const auto &KVP) {
+    checkSameImpl(KVP.second, Preorder);
+    return true;
+  });
+}
+
+void checkSame(const PerThreadCallsiteTrie &RCT, ArrayRef<Marker> Preorder) {
+  checkSameImpl(RCT.start(), Preorder);
+  ASSERT_TRUE(Preorder.empty());
+}
+
+TEST(PerThreadCallsiteTrieTest, Insert) {
+  PerThreadCallsiteTrie R;
+  uptr Stack1[]{4, 3, 2, 1};
+  R.insertStack(StackTrace(Stack1, 4));
+  checkSame(R, ArrayRef<Marker>({0, 1, 2, 3, 4, Marker::term()}));
```

Comment:
Do you need the explicit `ArrayRef<Marker>` constructor call?

Reply:
yes.